src/tree_util.rs

```diff
@@ -0,0 +1,196 @@
+use digest::Digest;
+
+/// The domain separator used for calculating leaf hashes
+const LEAF_HASH_PREFIX: &[u8] = &[0x00];
+
+/// The domain separator used for calculating parent hashes
+const PARENT_HASH_PREFIX: &[u8] = &[0x01];
+
+// We make opaque types for leaf and internal node indices so that we don't accidentally confuse
+// them in the math
+
+/// An index to a leaf of the tree
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub(crate) struct LeafIdx(u64);
```

Comment:
Naive question: Why not store this as a `usize` instead of `u64` to make both the `as_u64` and `as_usize` methods infallible?

I suppose this is to allow us to have a non-memory-backed tree that potentially stores up to `u64::MAX` nodes instead of the `usize::MAX` supported by `mem_backed_tree`?